text/0091-ci-upload-tokens.md

```diff
@@ -0,0 +1,273 @@
+still too high for many customers to make the necessary investment into getting source maps
+uploaded.  From a documentation writing and onboarding experience, it's also not clear with
+the current system how the experience can be improved.
+
+Additionally both Hybrid Cloud and Single Tenant would greatly benefit from automatically
+routing to the right URLs.  Today the documentation is very quiet about how to get this
+system to work on a single tenant installation and customers are often required to work
+with CS to get source maps working.
+
+# Technical Implementation
+
+The motivation is to add a new kind of token to Sentry which are fundamentally per-organization
+tokens, but with the ability to carry meta information that tools like sentry-cli can use to
+improve the user experience.  These org level tokens can be created by anyone in the org, they
+can be given additional restrictions, and they can carry meta information such as routing
+data.  For the purpose of this document they are called **structural tokens**.
+
+## Token Format
+
+The proposed token format is undecided so far.  The goals of the token align generally with
+both [Macaroons](http://macaroons.io/) and [Biscuit](https://www.biscuitsec.org).  Unfortunately
+the former standard has never seen much attention, and the latter is pretty new and not
+particularly proven.  Either system however permits adding additional restrictions to the
+token which make them a perfect choice for the use in our pipeline.  Biscuits however seem
+quite promising.  The idea of biscuits is that the token itself holds _facts_ which and
+can be further constraints via _checks_ and they are checked against a datalog inspired
+_policies_.
+
+One of the benefits of having the tokens carry this data is that the token alone has enough
+information available to route to a Sentry installation.  This means that `sentry-cli` or
+any other tool _just_ needs the token to even determine the host that the token should be
+sent against.  This benefit also applies to JWT or PASETO tokens which can be considered
+for this as well.  The RFC here thus proposes two potential options: A **Biscuit** token
+format and a **JWT** token format.
+
+A serialized token is added a custom prefix `sntrys_` (sentry structure) to make
+it possible to detect it by security scrapers.  Anyone handling such a token is
+required to check for the `sntrys_` prefix and disregard it before parsing it.  This
+can also be used by the client side to detect a structural token if the client is
+interested in extracting data from the token.
+
+## Token Facts
+
+We want to encode certain information into the tokens.  In Biscuit terms these are called
+_facts_.  The following facts exist:
+
+* `site`: references the target API URL that should be used.  A token will always have a
+  site in it and clients are not supposed to provide a fallback.  For instance this
+  would be `https://myorg.sentry.io/`.
```

Comment:
For hybrid cloud we won't be able to have API endpoints use `orgslug.sentry.io` without proxying through control silo. However, we could have structured tokens contain regional domains eg. `us.sentry.io`

Reply:
whys this different than org? why do we need this site component?

Reply:
For hybrid cloud the benefit is that you can (without a preflight) hit the right instance. Particularly for debug file uploads we do not want them to go via the central proxy at all times. For single tenant they all require different URLs anyways and today the docs are quite blind on what you have to reconfigure to point it to the right single tenant instance.